teerex/src/lib.rs

```diff
@@ -237,13 +258,42 @@ pub mod pallet {
 				sender_enclave.mr_enclave.encode() == shard_id.encode(),
 				<Error<T>>::WrongMrenclaveForShard
 			);
-			<WorkerForShard<T>>::insert(shard_id, sender_index);
-			log::debug!(
-				"Proposed sidechain block confirmed with shard {:?}, block hash {:?}",
-				shard_id,
-				block_hash
-			);
-			Self::deposit_event(Event::ProposedSidechainBlock(sender, block_hash));
+
+			let lenience = T::EarlyBlockProposalLenience::get();
+			let mut latest_block_header = <LatestSidechainBlockHeader<T>>::get(shard_id);
+			let latest_block_number = latest_block_header.block_number;
+			let block_number = block_header.block_number;
+
+			if block_number > Self::compare_block_number(latest_block_number, lenience)? {
```

Comment:
`compare_block_number` does not compare but add the numbers. But before changing this, I'd suggest a totally different  design altogether. Because it is hard to understand why the same function is called three times.. the differing 2nd input variable in only visible when looking at veery intensely.

Suggestion:

```rust
ensure!(block_number > latest_block_number , Error::<T>::OutdatedBlockNumber)
let MAX_ALLOWED_BLOCK_NUMBER = latest_block_number.checked_add(1).ok_or(Error::<T>::CheckedAddOverflow)?
ensure!(block_number < MAX_ALLOWED_BLOCK_NUMBER , Error::<T>::BlockNumberTooHigh)

let expected_block_number = latest_block_number.checked_add(1).ok_or(Error::<T>::CheckedAddOverflow)?
if block_number == expected_block_number {
 Self::finalize_sidechain_block(block_header)?
 Self::import_blocks_from_queue(...)?
} else {
 Self::import_to_queue(block_header)?;
}
Ok(())
```

This way the huge else if statement is also not necessary. What do you think?

Reply:
Ofc, the code won't compile. It's only a design suggestion.

Reply:
The naming of the functions and variables are not yet thought through. Fell free to adapt.

Reply:
Agreed that the function name is misleading. You could still have that function, called something with 'add' in it. And then assign it to variables with proper names (just as Bigna proposed). That way, you hide all the `checked_add` complexity (which contrary to Bigna's proposal I don't think belong in this top-level function), but still have readable code.

Reply:
I prefer to see 4 logically seperated cases. This is what the design relies on. The checks should done by the same method to prevent chaning one statement could introduce a gap or intersection of cases.

Reply:
But it's very hard to see which case leads to what (high cognitive complexity, however it's called). And I don't see why an else if statement would ensure that no intersection is happening? You could just change the check condition of one and there you go - intersection is there... it's just a matter of ordering which condition get checked first, but that's the case in my design proposal as well.

But I'll concede on that matter, as you guys have way more experience on that.